File: typescript/src/metadata.ts
import { decodeCompact } from "./scale.js";

const METADATA_MAGIC = 0x6174_656d;

export class MetadataError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MetadataError";
  }
}

export class ScaleError extends MetadataError {
  constructor(message: string) {
    super(message);
    this.name = "ScaleError";
  }
}

export class StorageNotFoundError extends MetadataError {
  constructor(message: string) {
    super(message);
    this.name = "StorageNotFoundError";
  }
}

export class FieldNotFoundError extends MetadataError {
  constructor(message: string) {
    super(message);
    this.name = "FieldNotFoundError";
  }
}

export class StorageValueTooShortError extends MetadataError {
  constructor(message: string) {
    super(message);
    this.name = "StorageValueTooShortError";
  }
}

class Reader {
  data: Uint8Array;
  pos: number;

  constructor(data: Uint8Array) {
    this.data = data;
    this.pos = 0;
  }

  read(n: number): Uint8Array {
    if (this.pos + n > this.data.length) {
      throw new ScaleError(
        `insufficient data: need ${n} bytes at ${this.pos}, have ${this.data.length - this.pos}`,
      );
    }
    const out = this.data.subarray(this.pos, this.pos + n);
    this.pos += n;
    return out;
  }

  readU8(): number {
    return this.read(1)[0];
  }

  readU32(): number {
    const b = this.read(4);
    return (b[0] | (b[1] << 8) | (b[2] << 16) | (b[3] << 24)) >>> 0;
  }

  readCompact(): number {
    const decoded = decodeCompact(this.data.subarray(this.pos));
    if (decoded === null) {
      throw new ScaleError("insufficient data for compact integer");
    }
    const [value, consumed] = decoded;
    this.pos += consumed;
    return Number(value);
  }

  readString(): string {
    const length = this.readCompact();
    const bytes = this.read(length);
    return new TextDecoder().decode(bytes);
  }

  readOptionString(): string | null {
    const tag = this.readU8();
    if (tag === 0) return null;
    if (tag === 1) return this.readString();
    throw new ScaleError(`invalid Option tag ${tag}`);
  }

  readVecString(): string[] {
    const n = this.readCompact();
    const out: string[] = [];
    for (let i = 0; i < n; i++) out.push(this.readString());
    return out;
  }

  readVecU8(): Uint8Array {
    const n = this.readCompact();
    return this.read(n);
  }

  skipStrings(): void {
    this.readVecString();
  }
}

interface Primitive {
  kind: "primitive";
  width: number;
  unsignedInt: boolean;
}

interface Composite {
  kind: "composite";
  fields: Array<[string, number]>;
}

interface ArrayShape {
  kind: "array";
  length: number;
  inner: number;
}

interface TupleShape {
  kind: "tuple";
  ids: number[];
}

interface VariantShape {
  kind: "variant";
  entries: Array<[number, string, string]>;
}

interface VariableShape {
  kind: "variable";
}

type TypeShape = Primitive | Composite | ArrayShape | TupleShape | VariantShape | VariableShape;

const VARIABLE: VariableShape = { kind: "variable" };

export interface StorageLayout {
  offset: number;
  width: number;
}

export function decodeUint(layout: StorageLayout, data: Uint8Array): bigint {
  const end = layout.offset + layout.width;
  if (data.length < end) {
    throw new StorageValueTooShortError(
      `storage value too short: need ${end} bytes, got ${data.length}`,
    );
  }
  let value = 0n;
  for (let i = 0; i < layout.width; i++) {
    value |= BigInt(data[layout.offset + i]) << BigInt(i * 8);
  }
  return value;
}

export interface ErrorEntry {
  pallet: string;
  variant: string;
  doc: string;
}

export class ErrorTable {
  byIdx: Map<string, ErrorEntry>;

  constructor() {
    this.byIdx = new Map();
  }

  humanize(palletIdx: number, errIdx: number): string | null {
    const entry = this.byIdx.get(`${palletIdx}:${errIdx}`);
    if (entry === undefined) return null;
    if (entry.doc === "") return `${entry.pallet}::${entry.variant}`;
    return `${entry.pallet}::${entry.variant}: ${entry.doc}`;
  }

  humanizeRpcError(raw: string): string {
    const payload = findAfterAny(raw, ["RPC error: ", "transaction failed: "]);
    if (payload !== null) {
      const jsonStr = trimToJson(payload);
      if (jsonStr !== null) {
        try {
          const parsed = JSON.parse(jsonStr);
          if (typeof parsed === "object" && parsed !== null) {
            const data = parsed.data;
            if (typeof data === "string") {
              const translated = this.maybeTranslateModule(data);
              return translated !== null ? translated : data;
            }
            const message = parsed.message;
            if (typeof message === "string") return message;
          }
        } catch {
          /* fallthrough */
        }
      }
    }
    const translated = this.maybeTranslateModule(raw);
    if (translated !== null) return translated;
    return raw;
  }

  private maybeTranslateModule(s: string): string | null {
    const start = s.indexOf("Module");
    if (start < 0) return null;
    const tail = s.slice(start);
    const idx = parseAfter(tail, "index:");
    const err = parseFirstByteAfter(tail, "error:");
    if (idx === null || err === null) return null;
    if (idx > 255 || err > 255) return null;
    return this.humanize(idx, err);
  }
}

function findAfterAny(s: string, needles: string[]): string | null {
  for (const n of needles) {
    const i = s.indexOf(n);
    if (i >= 0) return s.slice(i + n.length);
  }
  return null;
}

function trimToJson(s: string): string | null {
  const start = s.indexOf("{");
  if (start < 0) return null;
  let depth = 0;
  let inStr = false;
  let esc = false;
  for (let i = start; i < s.length; i++) {
    const b = s[i];
    if (esc) {
      esc = false;
      continue;
    }
    if (inStr && b === "\\") {
      esc = true;
      continue;
    }
    if (b === '"') {
      inStr = !inStr;
      continue;
    }
    if (inStr) continue;
    if (b === "{") depth++;
    else if (b === "}") {
      depth--;
      if (depth === 0) return s.slice(start, i + 1);
    }
  }
  return null;
}

function parseAfter(haystack: string, needle: string): number | null {
  const i = haystack.indexOf(needle);
  if (i < 0) return null;
  const rest = haystack.slice(i + needle.length);
  const m = rest.match(/^\s*(\d+)/);
  if (m === null) return null;
  return parseInt(m[1], 10);
}

function parseFirstByteAfter(haystack: string, needle: string): number | null {
  const i = haystack.indexOf(needle);
  if (i < 0) return null;
  const rest = haystack.slice(i + needle.length);
  const m = rest.match(/\[(\d+)/);
  if (m === null) return null;
  return parseInt(m[1], 10);
}

export class Metadata {
  private registry: TypeShape[];
  private storage: Map<string, number>;
  private calls: Map<string, [number, number]>;
  errors: ErrorTable;

  constructor(
    registry: TypeShape[],
    storage: Map<string, number>,
    calls: Map<string, [number, number]>,
    errors: ErrorTable,
  ) {
    this.registry = registry;
    this.storage = storage;
    this.calls = calls;
    this.errors = errors;
  }

  static fromRuntimeMetadata(data: Uint8Array): Metadata {
    const reader = new Reader(data);
    const magic = reader.readU32();
    if (magic !== METADATA_MAGIC) {
      throw new ScaleError(`metadata magic mismatch: 0x${magic.toString(16).padStart(8, "0")}`);
    }
    const version = reader.readU8();
    if (version !== 14) {
      throw new ScaleError(`metadata version ${version} unsupported (need V14)`);
    }
    const registry = readRegistry(reader);
    const { storage, pendingCalls, pendingErrors } = walkPallets(reader);
    const errors = buildErrorTable(registry, pendingErrors);
    const calls = resolveCallIndices(registry, pendingCalls);
    return new Metadata(registry, storage, calls, errors);
  }

  storageLayout(pallet: string, entry: string, fieldPath: string[]): StorageLayout {
    const valueTy = this.storage.get(`${pallet}.${entry}`);
    if (valueTy === undefined) {
      throw new StorageNotFoundError(`storage entry not found: ${pallet}.${entry}`);
    }
    let offset = 0;
    let current = valueTy;
    for (const fieldName of fieldPath) {
      const shape = this.typeAt(current);
      if (shape.kind !== "composite") {
        throw new MetadataError(`path traversal is not a composite at ${fieldName}`);
      }
      let found: number | null = null;
      for (const [name, ty] of shape.fields) {
        if (name === fieldName) {
          found = ty;
          break;
        }
        offset += this.byteSize(ty);
      }
      if (found === null) {
        throw new FieldNotFoundError(`field not found: ${fieldName}`);
      }
      current = found;
    }
    const shape = this.typeAt(current);
    if (shape.kind !== "primitive" || !shape.unsignedInt) {
      throw new MetadataError("storage_layout target is not an unsigned integer primitive");
    }
    return { offset, width: shape.width };
  }

  findCallIndex(pallet: string, call: string): [number, number] | null {
    const result = this.calls.get(`${pallet}.${call}`);
    return result === undefined ? null : result;
  }

  private typeAt(typeId: number): TypeShape {
    if (typeId < 0 || typeId >= this.registry.length) {
      throw new MetadataError(`type id ${typeId} missing from registry`);
    }
    return this.registry[typeId];
  }

  private byteSize(typeId: number): number {
    const shape = this.typeAt(typeId);
    if (shape.kind === "primitive") return shape.width;
    if (shape.kind === "composite") {
      let sum = 0;
      for (const [, ty] of shape.fields) sum += this.byteSize(ty);
      return sum;
    }
    if (shape.kind === "array") return shape.length * this.byteSize(shape.inner);
    if (shape.kind === "tuple") {
      let sum = 0;
      for (const t of shape.ids) sum += this.byteSize(t);
      return sum;
    }
    throw new MetadataError(`type id ${typeId} has variable width`);
  }
}

function readRegistry(reader: Reader): TypeShape[] {
  const n = reader.readCompact();
  const out: TypeShape[] = [];
  for (let expected = 0; expected < n; expected++) {
    const typeId = reader.readCompact();
    if (typeId !== expected) {
      throw new ScaleError(`non-sequential type id ${typeId} (expected ${expected})`);
    }
    reader.skipStrings();
    skipTypeParams(reader);
    out.push(readTypeDef(reader));
    reader.skipStrings();
  }
  return out;
}

function readTypeDef(reader: Reader): TypeShape {
  const tag = reader.readU8();
  if (tag === 0) return { kind: "composite", fields: readFields(reader) };
  if (tag === 1) {
    const n = reader.readCompact();
    const entries: Array<[number, string, string]> = [];
    for (let i = 0; i < n; i++) {
      const name = reader.readString();
      readFields(reader);
      const index = reader.readU8();
      const docs = reader.readVecString();
      let doc = "";
      for (const d of docs) {
        const trimmed = d.trim();
        if (trimmed !== "") {
          doc = trimmed;
          break;
        }
      }
      entries.push([index, name, doc]);
    }
    return { kind: "variant", entries };
  }
  if (tag === 2) {
    reader.readCompact();
    return VARIABLE;
  }
  if (tag === 3) {
    const length = reader.readU32();
    const inner = reader.readCompact();
    return { kind: "array", length, inner };
  }
  if (tag === 4) {
    const n = reader.readCompact();
    const ids: number[] = [];
    for (let i = 0; i < n; i++) ids.push(reader.readCompact());
    return { kind: "tuple", ids };
  }
  if (tag === 5) return primitiveShape(reader.readU8());
  if (tag === 6) {
    reader.readCompact();
    return VARIABLE;
  }
  if (tag === 7) {
    reader.readCompact();
    reader.readCompact();
    return VARIABLE;
  }
  throw new ScaleError(`unknown TypeDef tag ${tag}`);
}

function readFields(reader: Reader): Array<[string, number]> {
  const n = reader.readCompact();
  const out: Array<[string, number]> = [];
  for (let i = 0; i < n; i++) {
    const name = reader.readOptionString() ?? "";
    const ty = reader.readCompact();
    reader.readOptionString();
    reader.skipStrings();
    out.push([name, ty]);
  }
  return out;
}

function skipTypeParams(reader: Reader): void {
  const n = reader.readCompact();
  for (let i = 0; i < n; i++) {
    reader.readString();
    const tag = reader.readU8();
    if (tag === 0) continue;
    if (tag === 1) {
      reader.readCompact();
      continue;
    }
    throw new ScaleError(`invalid Option tag ${tag}`);
  }
}

function primitiveShape(tag: number): TypeShape {
  const table: Record<number, [number, boolean]> = {
    0: [1, false],
    1: [4, false],
    3: [1, true],
    4: [2, true],
    5: [4, true],
    6: [8, true],
    7: [16, true],
    8: [32, true],
    9: [1, false],
    10: [2, false],
    11: [4, false],
    12: [8, false],
    13: [16, false],
    14: [32, false],
  };
  if (tag === 2) return VARIABLE;
  if (!(tag in table)) throw new ScaleError(`unknown primitive tag ${tag}`);
  const [width, unsignedInt] = table[tag];
  return { kind: "primitive", width, unsignedInt };
}

interface PalletWalkResult {
  storage: Map<string, number>;
  pendingCalls: Array<[string, number, number]>;
  pendingErrors: Array<[string, number, number]>;
}

function walkPallets(reader: Reader): PalletWalkResult {
  const storage = new Map<string, number>();
  const pendingCalls: Array<[string, number, number]> = [];
  const pendingErrors: Array<[string, number, number]> = [];

  const n = reader.readCompact();
  for (let i = 0; i < n; i++) {
    const palletName = reader.readString();

    if (optionTag(reader)) {
      reader.readString();
      const entryCount = reader.readCompact();
      for (let j = 0; j < entryCount; j++) {
        const entryName = reader.readString();
        reader.readU8();
        const valueTy = readStorageEntryValueType(reader);
        storage.set(`${palletName}.${entryName}`, valueTy);
        reader.readVecU8();
        reader.skipStrings();
      }
    }

    const callsTy = optionTag(reader) ? reader.readCompact() : null;
    skipOptionalCompact(reader);

    const constCount = reader.readCompact();
    for (let j = 0; j < constCount; j++) {
      reader.readString();
      reader.readCompact();
      reader.readVecU8();
      reader.skipStrings();
    }

    const errorTy = optionTag(reader) ? reader.readCompact() : null;
    const palletIndex = reader.readU8();

    if (callsTy !== null) pendingCalls.push([palletName, palletIndex, callsTy]);
    if (errorTy !== null) pendingErrors.push([palletName, palletIndex, errorTy]);
  }

  return { storage, pendingCalls, pendingErrors };
}

function readStorageEntryValueType(reader: Reader): number {
  const tag = reader.readU8();
  if (tag === 0) return reader.readCompact();
  if (tag === 1) {
    const n = reader.readCompact();
    for (let i = 0; i < n; i++) reader.readU8();
    reader.readCompact();
    return reader.readCompact();
  }
  throw new ScaleError(`unknown StorageEntryType tag ${tag}`);
}

function optionTag(reader: Reader): boolean {
  const tag = reader.readU8();
  if (tag === 0) return false;
  if (tag === 1) return true;
  throw new ScaleError(`invalid Option tag ${tag}`);
}

function skipOptionalCompact(reader: Reader): void {
  if (optionTag(reader)) reader.readCompact();
}

function buildErrorTable(
  registry: TypeShape[],
  pending: Array<[string, number, number]>,
): ErrorTable {
  const table = new ErrorTable();
  for (const [palletName, palletIdx, errorTy] of pending) {
    if (errorTy < 0 || errorTy >= registry.length) continue;
    const shape = registry[errorTy];
    if (shape.kind !== "variant") continue;
    for (const [variantIdx, variantName, doc] of shape.entries) {
      table.byIdx.set(`${palletIdx}:${variantIdx}`, {
        pallet: palletName,
        variant: variantName,
        doc,
      });
    }
  }
  return table;
}

function resolveCallIndices(
  registry: TypeShape[],
  pending: Array<[string, number, number]>,
): Map<string, [number, number]> {
  const out = new Map<string, [number, number]>();
  for (const [palletName, palletIdx, callsTy] of pending) {
    if (callsTy < 0 || callsTy >= registry.length) continue;
    const shape = registry[callsTy];
    if (shape.kind !== "variant") continue;
    for (const [callIdx, callName] of shape.entries) {
      out.set(`${palletName}.${callName}`, [palletIdx, callIdx]);
    }
  }
  return out;
}

File: python/samp/metadata.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Optional, Union

from samp.scale import decode_compact

METADATA_MAGIC = 0x6174_656D


class MetadataError(Exception):
    pass


class ScaleError(MetadataError):
    pass


class StorageNotFoundError(MetadataError):
    pass


class FieldNotFoundError(MetadataError):
    pass


class StorageValueTooShortError(MetadataError):
    pass


class TypeIdMissingError(MetadataError):
    pass


class VariableWidthError(MetadataError):
    pass


class ShapeError(MetadataError):
    pass


class _Reader:
    def __init__(self, data: bytes) -> None:
        self.data = data
        self.pos = 0

    def read(self, n: int) -> bytes:
        if self.pos + n > len(self.data):
            raise ScaleError(
                f"insufficient data: need {n} bytes at {self.pos}, have {len(self.data) - self.pos}"
            )
        out = self.data[self.pos : self.pos + n]
        self.pos += n
        return out

    def read_u8(self) -> int:
        return self.read(1)[0]

    def read_u32(self) -> int:
        return int.from_bytes(self.read(4), "little")

    def read_compact(self) -> int:
        decoded = decode_compact(self.data[self.pos :])
        if decoded is None:
            raise ScaleError("insufficient data for compact integer")
        value, consumed = decoded
        self.pos += consumed
        return value

    def read_string(self) -> str:
        length = self.read_compact()
        return self.read(length).decode("utf-8")

    def read_option_string(self) -> Optional[str]:
        tag = self.read_u8()
        if tag == 0:
            return None
        if tag == 1:
            return self.read_string()
        raise ScaleError(f"invalid Option tag {tag}")

    def read_vec_string(self) -> list[str]:
        n = self.read_compact()
        return [self.read_string() for _ in range(n)]

    def read_vec_u8(self) -> bytes:
        n = self.read_compact()
        return self.read(n)

    def skip_strings(self) -> None:
        self.read_vec_string()


@dataclass(frozen=True)
class _Primitive:
    width: int
    unsigned_int: bool


@dataclass(frozen=True)
class _Composite:
    fields: tuple[tuple[str, int], ...]


@dataclass(frozen=True)
class _Array:
    length: int
    inner: int


@dataclass(frozen=True)
class _Tuple:
    ids: tuple[int, ...]


@dataclass(frozen=True)
class _Variant:
    entries: tuple[tuple[int, str, str], ...]


@dataclass(frozen=True)
class _Variable:
    pass


_VARIABLE = _Variable()

_TypeShape = Union[_Primitive, _Composite, _Array, _Tuple, _Variant, _Variable]


@dataclass(frozen=True)
class StorageLayout:
    offset: int
    width: int

    def decode_uint(self, data: bytes) -> int:
        end = self.offset + self.width
        if len(data) < end:
            raise StorageValueTooShortError(
                f"storage value too short: need {end} bytes, got {len(data)}"
            )
        return int.from_bytes(data[self.offset : end], "little")


@dataclass(frozen=True)
class ErrorEntry:
    pallet: str
    variant: str
    doc: str


@dataclass
class ErrorTable:
    by_idx: dict[tuple[int, int], ErrorEntry] = field(default_factory=dict)

    def humanize(self, pallet_idx: int, err_idx: int) -> Optional[str]:
        entry = self.by_idx.get((pallet_idx, err_idx))
        if entry is None:
            return None
        if not entry.doc:
            return f"{entry.pallet}::{entry.variant}"
        return f"{entry.pallet}::{entry.variant}: {entry.doc}"

    def humanize_rpc_error(self, raw: str) -> str:
        payload = _find_after_any(raw, ["RPC error: ", "transaction failed: "])
        if payload is not None:
            json_str = _trim_to_json(payload)
            if json_str is not None:
                try:
                    parsed = json.loads(json_str)
                except json.JSONDecodeError:
                    parsed = None
                if isinstance(parsed, dict):
                    data = parsed.get("data")
                    if isinstance(data, str):
                        translated = self._maybe_translate_module(data)
                        return translated if translated is not None else data
                    message = parsed.get("message")
                    if isinstance(message, str):
                        return message
        translated = self._maybe_translate_module(raw)
        if translated is not None:
            return translated
        return raw

    def _maybe_translate_module(self, s: str) -> Optional[str]:
        start = s.find("Module")
        if start < 0:
            return None
        tail = s[start:]
        idx = _parse_after(tail, "index:")
        err = _parse_first_byte_after(tail, "error:")
        if idx is None or err is None:
            return None
        if idx > 255 or err > 255:
            return None
        return self.humanize(idx, err)


def _find_after_any(s: str, needles: list[str]) -> Optional[str]:
    for n in needles:
        i = s.find(n)
        if i >= 0:
            return s[i + len(n) :]
    return None


def _trim_to_json(s: str) -> Optional[str]:
    start = s.find("{")
    if start < 0:
        return None
    depth = 0
    in_str = False
    esc = False
    for i in range(start, len(s)):
        b = s[i]
        if esc:
            esc = False
            continue
        if in_str and b == "\\":
            esc = True
            continue
        if b == '"':
            in_str = not in_str
            continue
        if in_str:
            continue
        if b == "{":
            depth += 1
        elif b == "}":
            depth -= 1
            if depth == 0:
                return s[start : i + 1]
    return None


_DIGITS_AFTER = re.compile(r"\s*(\d+)")
_DIGITS_BRACKET = re.compile(r"\[(\d+)")


def _parse_after(haystack: str, needle: str) -> Optional[int]:
    i = haystack.find(needle)
    if i < 0:
        return None
    rest = haystack[i + len(needle) :]
    m = _DIGITS_AFTER.match(rest)
    if m is None:
        return None
    return int(m.group(1))


def _parse_first_byte_after(haystack: str, needle: str) -> Optional[int]:
    i = haystack.find(needle)
    if i < 0:
        return None
    rest = haystack[i + len(needle) :]
    m = _DIGITS_BRACKET.search(rest)
    if m is None:
        return None
    return int(m.group(1))


@dataclass(frozen=True)
class Metadata:
    registry: tuple[_TypeShape, ...]
    storage: dict[tuple[str, str], int]
    calls: dict[tuple[str, str], tuple[int, int]]
    errors: ErrorTable

    @classmethod
    def from_runtime_metadata(cls, data: bytes) -> "Metadata":
        reader = _Reader(data)
        magic = reader.read_u32()
        if magic != METADATA_MAGIC:
            raise ScaleError(f"metadata magic mismatch: 0x{magic:08x}")
        version = reader.read_u8()
        if version != 14:
            raise ScaleError(f"metadata version {version} unsupported (need V14)")

        registry = _read_registry(reader)
        storage, pending_calls, pending_errors = _walk_pallets(reader)
        errors = _build_error_table(registry, pending_errors)
        calls = _resolve_call_indices(registry, pending_calls)

        return cls(
            registry=tuple(registry),
            storage=storage,
            calls=calls,
            errors=errors,
        )

    def storage_layout(
        self,
        pallet: str,
        entry: str,
        field_path: list[str],
    ) -> StorageLayout:
        value_ty = self.storage.get((pallet, entry))
        if value_ty is None:
            raise StorageNotFoundError(f"storage entry not found: {pallet}.{entry}")
        offset = 0
        current = value_ty
        for field_name in field_path:
            shape = _type_at(self.registry, current)
            if not isinstance(shape, _Composite):
                raise ShapeError(f"path traversal is not a composite at {field_name}")
            found: Optional[int] = None
            for name, ty in shape.fields:
                if name == field_name:
                    found = ty
                    break
                offset += _byte_size(self.registry, ty)
            if found is None:
                raise FieldNotFoundError(f"field not found: {field_name}")
            current = found
        shape = _type_at(self.registry, current)
        if not isinstance(shape, _Primitive) or not shape.unsigned_int:
            raise ShapeError("storage_layout target is not an unsigned integer primitive")
        return StorageLayout(offset=offset, width=shape.width)

    def find_call_index(self, pallet: str, call: str) -> Optional[tuple[int, int]]:
        return self.calls.get((pallet, call))


def _type_at(registry: tuple[_TypeShape, ...], type_id: int) -> _TypeShape:
    if type_id < 0 or type_id >= len(registry):
        raise TypeIdMissingError(f"type id {type_id} missing from registry")
    return registry[type_id]


def _byte_size(registry: tuple[_TypeShape, ...], type_id: int) -> int:
    shape = _type_at(registry, type_id)
    if isinstance(shape, _Primitive):
        return shape.width
    if isinstance(shape, _Composite):
        return sum(_byte_size(registry, ty) for _, ty in shape.fields)
    if isinstance(shape, _Array):
        return shape.length * _byte_size(registry, shape.inner)
    if isinstance(shape, _Tuple):
        return sum(_byte_size(registry, t) for t in shape.ids)
    raise VariableWidthError(f"type id {type_id} has variable width")


def _read_registry(reader: _Reader) -> tuple[_TypeShape, ...]:
    n = reader.read_compact()
    out: list[_TypeShape] = []
    for expected in range(n):
        type_id = reader.read_compact()
        if type_id != expected:
            raise ScaleError(f"non-sequential type id {type_id} (expected {expected})")
        reader.skip_strings()
        _skip_type_params(reader)
        out.append(_read_type_def(reader))
        reader.skip_strings()
    return tuple(out)


def _read_type_def(reader: _Reader) -> _TypeShape:
    tag = reader.read_u8()
    if tag == 0:
        return _Composite(fields=tuple(_read_fields(reader)))
    if tag == 1:
        n = reader.read_compact()
        entries: list[tuple[int, str, str]] = []
        for _ in range(n):
            name = reader.read_string()
            _read_fields(reader)
            index = reader.read_u8()
            docs = reader.read_vec_string()
            doc = next((d.strip() for d in docs if d.strip()), "")
            entries.append((index, name, doc))
        return _Variant(entries=tuple(entries))
    if tag == 2:
        reader.read_compact()
        return _VARIABLE
    if tag == 3:
        length = reader.read_u32()
        inner = reader.read_compact()
        return _Array(length=length, inner=inner)
    if tag == 4:
        n = reader.read_compact()
        ids = tuple(reader.read_compact() for _ in range(n))
        return _Tuple(ids=ids)
    if tag == 5:
        return _primitive_shape(reader.read_u8())
    if tag == 6:
        reader.read_compact()
        return _VARIABLE
    if tag == 7:
        reader.read_compact()
        reader.read_compact()
        return _VARIABLE
    raise ScaleError(f"unknown TypeDef tag {tag}")


def _read_fields(reader: _Reader) -> list[tuple[str, int]]:
    n = reader.read_compact()
    out: list[tuple[str, int]] = []
    for _ in range(n):
        name = reader.read_option_string() or ""
        ty = reader.read_compact()
        reader.read_option_string()
        reader.skip_strings()
        out.append((name, ty))
    return out


def _skip_type_params(reader: _Reader) -> None:
    n = reader.read_compact()
    for _ in range(n):
        reader.read_string()
        tag = reader.read_u8()
        if tag == 0:
            continue
        if tag == 1:
            reader.read_compact()
            continue
        raise ScaleError(f"invalid Option tag {tag}")


def _primitive_shape(tag: int) -> _TypeShape:
    table: dict[int, tuple[int, bool]] = {
        0: (1, False),
        1: (4, False),
        3: (1, True),
        4: (2, True),
        5: (4, True),
        6: (8, True),
        7: (16, True),
        8: (32, True),
        9: (1, False),
        10: (2, False),
        11: (4, False),
        12: (8, False),
        13: (16, False),
        14: (32, False),
    }
    if tag == 2:
        return _VARIABLE
    if tag not in table:
        raise ScaleError(f"unknown primitive tag {tag}")
    width, unsigned_int = table[tag]
    return _Primitive(width=width, unsigned_int=unsigned_int)


def _walk_pallets(
    reader: _Reader,
) -> tuple[
    dict[tuple[str, str], int],
    list[tuple[str, int, int]],
    list[tuple[str, int, int]],
]:
    storage: dict[tuple[str, str], int] = {}
    pending_calls: list[tuple[str, int, int]] = []
    pending_errors: list[tuple[str, int, int]] = []

    n = reader.read_compact()
    for _ in range(n):
        pallet_name = reader.read_string()

        if _option_tag(reader):
            reader.read_string()
            entry_count = reader.read_compact()
            for _ in range(entry_count):
                entry_name = reader.read_string()
                reader.read_u8()
                value_ty = _read_storage_entry_value_type(reader)
                storage[(pallet_name, entry_name)] = value_ty
                reader.read_vec_u8()
                reader.skip_strings()

        calls_ty = reader.read_compact() if _option_tag(reader) else None
        _skip_optional_compact(reader)

        const_count = reader.read_compact()
        for _ in range(const_count):
            reader.read_string()
            reader.read_compact()
            reader.read_vec_u8()
            reader.skip_strings()

        error_ty = reader.read_compact() if _option_tag(reader) else None
        pallet_index = reader.read_u8()

        if calls_ty is not None:
            pending_calls.append((pallet_name, pallet_index, calls_ty))
        if error_ty is not None:
            pending_errors.append((pallet_name, pallet_index, error_ty))

    return storage, pending_calls, pending_errors


def _read_storage_entry_value_type(reader: _Reader) -> int:
    tag = reader.read_u8()
    if tag == 0:
        return reader.read_compact()
    if tag == 1:
        n = reader.read_compact()
        for _ in range(n):
            reader.read_u8()
        reader.read_compact()
        return reader.read_compact()
    raise ScaleError(f"unknown StorageEntryType tag {tag}")


def _option_tag(reader: _Reader) -> bool:
    tag = reader.read_u8()
    if tag == 0:
        return False
    if tag == 1:
        return True
    raise ScaleError(f"invalid Option tag {tag}")


def _skip_optional_compact(reader: _Reader) -> None:
    if _option_tag(reader):
        reader.read_compact()


def _build_error_table(
    registry: tuple[_TypeShape, ...], pending: list[tuple[str, int, int]]
) -> ErrorTable:
    table = ErrorTable()
    for pallet_name, pallet_idx, error_ty in pending:
        if error_ty < 0 or error_ty >= len(registry):
            continue
        shape = registry[error_ty]
        if not isinstance(shape, _Variant):
            continue
        for variant_idx, variant_name, doc in shape.entries:
            table.by_idx[(pallet_idx, variant_idx)] = ErrorEntry(
                pallet=pallet_name,
                variant=variant_name,
                doc=doc,
            )
    return table


def _resolve_call_indices(
    registry: tuple[_TypeShape, ...], pending: list[tuple[str, int, int]]
) -> dict[tuple[str, str], tuple[int, int]]:
    out: dict[tuple[str, str], tuple[int, int]] = {}
    for pallet_name, pallet_idx, calls_ty in pending:
        if calls_ty < 0 or calls_ty >= len(registry):
            continue
        shape = registry[calls_ty]
        if not isinstance(shape, _Variant):
            continue
        for entry in shape.entries:
            call_idx, call_name = entry[0], entry[1]
            out[(pallet_name, call_name)] = (pallet_idx, call_idx)
    return out
